ntoskrnl/include/internal/i386/mm.h

```diff
@@ -128,21 +117,68 @@
 /* On x86, these two are the same */
 #define MI_WRITE_VALID_PPE MI_WRITE_VALID_PTE
 
-/* Convert an address to a corresponding PTE */
-#define MiAddressToPte(x) \
-    ((PMMPTE)(((((ULONG)(x)) >> 12) << 2) + PTE_BASE))
+/*  Translating virtual addresses to physical addresses
+        (See: "Intel� 64 and IA-32 Architectures Software Developer�s Manual
+              Volume 3A: System Programming Guide, Part 1, CHAPTER 4 PAGING")
+    Page directory (PD) and Page table (PT) definitions
+    Page directory entry (PDE) and Page table entry (PTE) definitions
+*/
+
+/* Maximum number of page directories pages */
+#ifndef _X86PAE_
+#define PD_COUNT 1        /* Only one page directory page */
+#else
+#define PD_COUNT (1 << 2) /* The two most significant bits in the VA */
+#endif
+
+/* The number of PTEs on one page of the PT */
+#define PTE_PER_PAGE (PAGE_SIZE / sizeof(MMPTE))
+
+/* The number of PDEs on one page of the PD */
+#define PDE_PER_PAGE (PAGE_SIZE / sizeof(MMPDE))
+
+/* Maximum number of PDEs */
+#define PDE_PER_SYSTEM (PD_COUNT * PDE_PER_PAGE)
+
+/* TODO: It seems this variable is not needed for x86 */
+#define PPE_PER_PAGE 1
+
+/* Maximum number of pages for 4 GB of virtual space */
+#define MI_MAX_PAGES ((1ull << 32) / PAGE_SIZE)
+
+/* Base addresses for page tables */
+#define PTE_BASE (ULONG_PTR)0xC0000000
+#define PTE_TOP  (ULONG_PTR)(PTE_BASE + (MI_MAX_PAGES * sizeof(MMPTE)) - 1)
+#define PTE_MASK (PTE_TOP - PTE_BASE)
+
+#define MI_SYSTEM_PTE_BASE (PVOID)MiAddressToPte(NULL)
+
+/* Base addreses for page directories */
+#define PDE_BASE (ULONG_PTR)MiAddressToPte(PTE_BASE)
```

Comment:
Using MiAddressToPte for defining a PDE is confusing. Either MiAddressToPde(NULL) or MiPteToPde(PTE_BASE)

Reply:
I agree, this looks clearer.